processing/src/main/java/io/druid/query/groupby/epinephelinae/EpiGroupByMergingQueryRunner.java

```diff
@@ -0,0 +1,645 @@
+      GroupByQueryConfig config,
+      ExecutorService exec,
+      QueryWatcher queryWatcher,
+      Iterable<QueryRunner<Row>> queryables,
+      int concurrencyHint,
+      BlockingPool<ByteBuffer> mergeBufferPool,
+      ObjectMapper spillMapper
+  )
+  {
+    this.config = config;
+    this.exec = MoreExecutors.listeningDecorator(exec);
+    this.queryWatcher = queryWatcher;
+    this.queryables = Iterables.unmodifiableIterable(Iterables.filter(queryables, Predicates.notNull()));
+    this.concurrencyHint = concurrencyHint;
+    this.mergeBufferPool = mergeBufferPool;
+    this.spillMapper = spillMapper;
+  }
+
+  @Override
+  public Sequence<Row> run(final Query queryParam, final Map responseContext)
+  {
+    final GroupByQuery query = (GroupByQuery) queryParam;
+
+    // CTX_KEY_MERGE_RUNNERS_USING_CHAINED_EXECUTION is here because realtime servers use nested mergeRunners calls
+    // (one for the entire query and one for each sink). We only want the outer call to actually do merging with a
+    // merge buffer, otherwise the query will allocate too many merge buffers. This is potentially sub-optimal as it
+    // will involve materializing the results for each sink before starting to feed them into the outer merge buffer.
+    // I'm not sure of a better way to do this without tweaking how realtime servers do queries.
+    final boolean forceChainedExecution = query.getContextBoolean(
+        CTX_KEY_MERGE_RUNNERS_USING_CHAINED_EXECUTION,
+        false
+    );
+    final GroupByQuery queryForRunners = query.withOverriddenContext(
+        ImmutableMap.<String, Object>of(CTX_KEY_MERGE_RUNNERS_USING_CHAINED_EXECUTION, true)
+    );
+
+    if (BaseQuery.getContextBySegment(query, false) || forceChainedExecution) {
+      return new ChainedExecutionQueryRunner(exec, queryWatcher, queryables).run(query, responseContext);
+    }
+
+    final AggregatorFactory[] combiningAggregatorFactories = new AggregatorFactory[query.getAggregatorSpecs().size()];
+    for (int i = 0; i < query.getAggregatorSpecs().size(); i++) {
+      combiningAggregatorFactories[i] = query.getAggregatorSpecs().get(i).getCombiningFactory();
+    }
+
+    final GroupByMergingKeySerdeFactory keySerdeFactory = new GroupByMergingKeySerdeFactory(
+        query.getDimensions().size(),
+        config.getMaxMergingDictionarySize() / concurrencyHint
+    );
+    final GroupByMergingColumnSelectorFactory columnSelectorFactory = new GroupByMergingColumnSelectorFactory();
+
+    final File temporaryStorageDirectory = new File(
+        System.getProperty("java.io.tmpdir"),
+        String.format("druid-groupBy-%s_%s", UUID.randomUUID(), query.getId())
+    );
+
+    final LimitedTemporaryStorage temporaryStorage = new LimitedTemporaryStorage(
+        temporaryStorageDirectory,
+        config.getMaxOnDiskStorage()
+    );
+
+    final ResourceHolder<ByteBuffer> mergeBufferHolder;
+
+    try {
+      // TODO(gianm): Should have a timeout here
```

Comment:
I think, we should use query timeout here. Sometimes broker can crash leaving these child queries orphaned.